### ISSUE TITLE
message headers failure

### ISSUE THREAD

**Issue by avivcortica** [vert-x3/vertx-rabbitmq-client#30]
Whenever I receive a message that contains a headers values, and i apply `includeProperties`, the framework [places them](https://github.com/vert-x3/vertx-rabbitmq-client/blob/97ef8a62917d2b33579bd289d27c18e878320c99/src/main/java/io/vertx/rabbitmq/impl/ConsumerHandler.java#L51) in the `JsonObject`, which causes the `JsonObject` serialization to fail with message:

```
io.vertx.core.json.EncodeException: Failed to encode as JSON: No serializer found for class java.io.DataInputStream and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) ) (through reference chain: java.util.LinkedHashMap["properties"]->java.util.LinkedHashMap["headers"]->java.util.HashMap["contentType"]->com.rabbitmq.client.impl.ByteArrayLongString["stream"])
```

This means that the message then cannot be sent on the event bus and cannot be consumed.
Am i doing something wrong? Or is this indeed an issue?

For me, in order to reproduce - i just added headers to the message.

Thanks.

**Comment by poiuytrez** [comment]
Would you be able to share a small reproducer on Github? Thank you !

**Comment by avivcortica** [comment]
Based on [this project](https://github.com/vert-x3/vertx-examples/tree/master/maven-simplest), i've added this to `pom.xml`:

```
<dependency>
      <groupId>io.vertx</groupId>
      <artifactId>vertx-rabbitmq-client</artifactId>
      <version>3.3.3</version>
    </dependency>
```

and altered `HelloWorldEmbedded` class to:

```
public class HelloWorldEmbedded {

    public static void main(String[] args) {
        Vertx vertx = Vertx.vertx();
        JsonObject config = new JsonObject();
        config.put("user", "guest");
        config.put("password", "guest");
        config.put("includeProperties", true);
        RabbitMQClient client = RabbitMQClient.create(vertx, config);
        client.start(startResult -> {
            if (!startResult.succeeded()) {
                throw new RuntimeException("failed 1", startResult.cause());
            }
            client.exchangeDeclare("exchange_name", "topic", true, false, exchangeDeclareResult -> {
                if (!exchangeDeclareResult.succeeded()) {
                    throw new RuntimeException("failed 2", exchangeDeclareResult.cause());
                }
                client.queueDeclare("queue_name", true, false, false, declareResult -> {
                    if (!declareResult.succeeded()) {
                        throw new RuntimeException("failed 3", declareResult.cause());
                    }
                    client.queueBind("queue_name", "exchange_name", "routing_key", bindResult -> {
                        if (!bindResult.succeeded()) {
                            throw new RuntimeException("failed 4", bindResult.cause());
                        }
                        vertx.eventBus().consumer("event_bus_address", message -> {
                            // not reaching this point
                            throw new RuntimeException("reached");
                        });
                        client.basicConsume("queue_name", "event_bus_address", consumeResult -> {
                            if (!consumeResult.succeeded()) {
                                throw new RuntimeException("failed 5", consumeResult.cause());
                            }
                        });
                    });
                });
            });
        });
    }

}
```

then i opened the Rabbit web management client, clicked `queues` button, clicked `queue_name` queue, went down to `Publish message`, at first i put `{}` in the payload and clicked `Publish message`, then the code reached the `// not reaching this point` block as originally expected.
Then when i add any header with any value, the serialization fails, causing the thread not to reach the consumer block, and the message to remain unacked.
Notice the `config.put("includeProperties", true)` above, causing the change.

**Comment by avivcortica** [comment]
BTW i actually use `3.3.4-SNAPSHOT` and the issue also occurs there.

**Comment by poiuytrez** [comment]
Thank you for the reproducer. We currently do not have any bandwidth to dig but we now have all the required informations.

**Comment by thibaultleouay** [comment]
Here's a quick fix for it, 

We should spend more time to use spi and a message codec

**Comment by Sammers21** [comment]
@vietj, has https://github.com/vert-x3/vertx-rabbitmq-client/pull/45 solved the problem?

**Comment by vietj** [comment]
I don't know we should try the reproducer

**Comment by vietj** [comment]
can you investigate @Sammers21 ?

**Comment by Sammers21** [comment]
@vietj , sure

**Comment by Sammers21** [comment]
@vietj, with new consumer API use, will not face the issue, can we close this?

**Comment by vietj** [comment]
do you mean the bug is fixed ? if not I think we should still try to fix it, even if the api is deprecated, unless it's impossible to fix by nature

**Comment by m1schka** [comment]
For me, the bug is not fixed (on 3.5.2 after #45 ). specifically if the value of the header is another object like the `x-death` object, when the original message was dead lettered and later put back into the queue. I guess the time is causing the issue here

Here a screenshow how the headers look like in the management console
<img width="302" alt="screen shot 2018-07-10 at 14 36 33" src="https://user-images.githubusercontent.com/1403301/42510357-f04a3e88-844e-11e8-848b-68b19aa33038.png">

BTW: what is the new consumer api you're talking about?

**Comment by jacek-wojciechowski-dell** [comment]
I have exactly the same problem as @m1schka described with x-death headers.
Therefore, I cannot use message headers and needed to invent custom message properties.

But would be really nice to have a fix for that.

**Comment by jonjomckay** [comment]
@vietj I've fixed the issue in #88 if you wouldn't mind taking a look?

**Comment by YaytayAtWork** [comment]
Can this issue be closed now?

**Comment by Yaytay** [comment]
Closing this as the includeProperties value is never read from the options!
